MUZIMA 462

samuelowino opened the pull request:
Added getJSONArrayFromObject() which processes the JSONObject from payload to a JSONArray for Processing attributes.

samuelowino committed:
Added getJSONArrayFromObject() 

Added getJSONArrayFromObject() which processes the JSONObject from payload to a JSONArray for Processing attributes

mssavai reviewed (commented):
Hey @samuelowino please see more description of the requirement on ticket and updte the PR before we can merge. Thanks!

samuelowino committed:
MUZIMA-461

Added subrotine to determine if JSON subtree is a JSONArray or JSONObject, to JsonUtils, this is intended to facilitate processing of the new payload structures [patient.personaddress] defined in muzima-461 design notes.

samuelowino commented:
Have a look at the modifications made on JsonUtils.java file and the unit tests.Awaiting guidance

mssavai commented on api/src/test/java/org/openmrs/module/muzima/handler/JsonUtilsTest.java:
```diff
@@ -0,0 +1,43 @@
+                "                        }\n" +
+                "                ]";
+
+        Object testPath = JsonUtils.readAsObject(payload,"$['patient.personaddress']['patient.countyDistrict']");
```
This won't work because you are giving invalid path to the payload. I'd expect that you do JsonUtils.readAsObject(payload,"$['patient.personaddress']"), then convert the returned object into a Json Array and iterate through in order to access values such as countyDistrict inside the array

samuelowino committed:
muzma-461--muzima-Made modifications to etPatientAddressesFromPayload() as improvements for a dynamic JSON processing strategy.

samuelowino commented:
Did not notice that there was already a test class in place, will move the test to that test class, Also please look at new changes made to JsonRegistrationQueuDataHandler.

mssavai commented on api/src/main/java/org/openmrs/module/muzima/handler/JsonRegistrationQueueDataHandler.java:
```diff
@@ -13,19 +13,16 @@
-import org.openmrs.PersonAttributeType;
-import org.openmrs.PersonName;
+import org.javarosa.core.services.Logger;
+import org.openmrs.*;
```
Could you avoid wildcard (*) imports? Your IDE should probably have a setting to disable automatically resorting to wildcard

samuelowino replied:
I am using Inteliji IDEA I believe there is a setting that can remove wildcard imports. I understand the confusion it can bring with Object mismatch and similar class names from different packages.

mssavai commented on api/src/main/java/org/openmrs/module/muzima/handler/JsonRegistrationQueueDataHandler.java:
```diff
@@ -275,35 +272,81 @@ private String getPatientUuidFromPayload(){
             //process as JSONArray
-            String county = JsonUtils.readAsString(payload, "$['patient']['patient.county']");
-            patientAddress.setStateProvince(county);
+            PersonAddress firstSetPatientAddress = new PersonAddress();
```
If JSONArray, there can be more than two addresses, you will therefore need to amend so as to accommodate a JSONArray of any length

samuelowino replied:
I also guessed the payload could contain more than two patient.personaddress objects or nodes, unfortunately, I limited myself to the example payloads in the notes, will make the necessary adjustments.

mssavai reviewed (commented):
Please see comments. Thanks

samuelowino committed:
Close to final modifications for patient.personaddress -MUZIMA-461

Added method to process nodes - $['patient']['patient.personaddress^n']. Also made modifications to the sections that process plain JSONObject and JSONArray instances of patient.personaddress.

samuelowino committed:
Begun work on personatrributes generic payload processing SEE line 383-419

The code refactoring so far should handle paload example 1 and 2. Carrying out adjustments for examples 3 i.e. patient.personattribute^n

samuelowino committed:
Made adjustments to accomodate examples 3 of the persnattribute payload structure. MUZIMA-461

samuelowino committed:
Patient Identifier edits init MUZIMA-461 subtask

samuelowino committed:
Completed processing subroutine for patientIdentifiers MUZIMA-461.

mssavai committed:
MUZIMA-536: Add ability to create settings

mssavai committed:
MUZIMA-536: more updates

mssavai committed:
MUZIMA-539: support 2.0.x and older Openmrs versions

mssavai committed:
MUZIMA-536: more updates for muzima settings

mssavai committed:
Merge branch 'MUZIMA-539' of https://github.com/mssavai/openmrs-module-muzimacore into MUZIMA-536

samuelowino committed:
Revaluated JsonProcessing facility, optimized JsonPath, tested and synced JsonUtils subrouitines

samuelowino committed:
Revaluated JsonProcessing facility, optimized JsonPath, tested and synced JsonUtils subrouitines

samuelowino commented:
Kindly preview new changes, noticed a hickup with JsonUtils.readAsObject method apparently it return a LinkedHashMap that needs to parsed to JSONObject, I have refactored my code to sync with this. 
Please let me know how this problem was hundled or if my testing was wrong, when I test that payload processing using the readAsObject method I got this erro **cannot parse java.util.LinedHashMap to net.minidev.JSONObject**

>

samuelowino committed:
Fixed wrong json node adjustments made in the previouse commit

samuelowino committed:
Merge branch 'master' of https://github.com/muzima/openmrs-module-muzimacore

mssavai committed:
Update supported OpenMRS version in REST resources

mssavai committed:
MUZIMA-536: add settings REST resource

mssavai committed:
Merge https://github.com/samuelowino/openmrs-module-muzimacore into MUZIMA-536

mssavai committed:
MUZIMA-536: Settings & more updates

mssavai committed:
MUZIMA-462: Updates for demographics update handler & refactoring

samuelowino committed:
changed access modifier for setPatientAddress- method for temporary unit test

samuelowino committed:
changed access modifier for setPatientAddress- method for temporary unit test

mssavai committed:
MUZIMA-461: Create new handler and revert registration handler for backward compatibility

samuelowino committed:
Created handler testing context with no autowired beans yet, also wrote the getDiscriminator test case

samuelowino committed:
Developed JsonUtils tests, for the JsonUtils handler util

samuelowino committed:
Added Logging support to the JsonUtilsTest, Completed remaining test methods, tests are in Nightly Builds, not executed and refactore yet

samuelowino commented:
The old JsonRegistration QueueDataHandler class went crazy when I merged your changes, there were some unidentified attributes that made my IDEA go red something like 'HEAD'  and other characters had to comment the whole source file.

When you can check it out commit id 4ab14d1

samuelowino commented:
Alson have you tried testing payload processing on muzima 2.0.0 android app with a build from MUZIMA-536 branch if you have, how did the server side respond... really need a testing device.

samuelowino committed:
Completely refactored JsonUtilsTest case methods, 12 of 15 tests pased on the run done at the time of this commit, one method  writeAsBoolean required production code refatoring, the remaing two tests writeAsDateTime and ReadAsDateTime require ISO80U..Date scrutiny

samuelowino committed:
Added and modified testing resources for faster test bootsrapping, added test-db config to test session factory and new generic payload sample from MUZIMA-461/462 design notes

samuelowino committed:
Refactored TestContextConfiguration facility for testing db bean and model objects
